mmcmds.c

```diff
@@ -4936,9 +4937,9 @@ void verifyMarkup(vstring labelMatch,
       continue;
     }
 
-    let(&str1, ""); /* Prevent string stack buildup/overflow in left() below */
     /* Look for "ax-*" axioms */
     if (strcmp("ax-", left(g_Statement[stmtNum].labelName, 3))) {
+      freeTempAlloc(); /* Prevent string stack buildup/overflow in left() */
```

Comment:
What is going on here?

Reply:
The call to `left` produces a temporary. Usually such temporaries are used in an argument to `let` so they are cleaned up thereafter, but in this case it is an argument to `strcmp` which does no such cleanup. This is not normally an issue since there are `let`s all over the place, but in this particular case there is the possibility that we get to L4941, allocate the temporary, and then `continue` at L4943, leaving the temp uncleared. If this happens for many loop iterations in a row then we can overflow the temp buffer.

The issue only happens in this branch: on previous `continue`s in this loop no temporary has been created, and after this if statement there is a `let` which will clean up the temporary.

The `let(&str1, "")` command is otherwise useless; `str1` is not in use at this point in the program. (I find it clearer to not randomly set variables that aren't in use.)

Reply:
The problem is that freeTempAlloc not only frees the temporary created by left, but other temporaries too.  Can you tell without thinking that str2, defined in line 4852:
let(&str2, cat(",", edit(g_Statement[stmtNum].labelName, 32/*uppercase*/),
        ",", NULL));
is not reused later?
While str2 is redefined here short after (so this particular case does not hurt), I had to make sure no variables are used ACROSS this loop.  And how do you know someone adds one in future, not knowing it is destroyed here?  This technique is inherently unsafe.

At least put a 'memory barrier' in form of a comment:
/* do not use temporary variables across this line */
in front of this loop.  Then there is at least an assertion to fulfil.

Reply:
> The problem is that freeTempAlloc not only frees the temporaries created by left, but other temporaries too. Can you tell without thinking that str2, defined in line 4852: let(&str2, cat(",", edit(g_Statement[stmtNum].labelName, 32/_uppercase_/), ",", NULL)); is not reused later?

Yes, I can tell this easily. `str2` is not a temporary (a `temp_vstring`), it is a regular `vstring`. Temporaries *must not* appear on the left of a `let` statement, and they must be cleaned up at the end of the scope (usually, the end of the function). They are not subject to cleanup by `let` statements.

A temporary would be something like the return value of `edit()` in that example. These are almost never stored in variables, they only exist as subexpressions (C temporaries), and they usually only last for the duration of one line of code, so it is generally pretty easy to see their scope.

As I mention in the other comment, `g_startTempAllocStack` is almost always 0, except inside the implementation of some vstring functions. There are no "local temporary scopes" in the main part of the code.